File: leetcode/677.map-sum-pairs.py
class TrieNode(object):
    def __init__(self):
        self.children = {}
        self.is_word = False
        self.sum = 0


class MapSum(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def insert(self, key, val):
        """
        :type key: str
        :type val: int
        :rtype: void
        """
        node = self.root
        for c in key:
            if c not in node.children:
                node.children[c] = TrieNode()

            node = node.children[c]

        node.is_word = True
        node.sum = val
        

    def sum(self, prefix):
        """
        :type prefix: str
        :rtype: int
        """
        node = self.root
        for c in prefix:
            if c not in node.children:
                return 0

            node = node.children[c]

        res = 0
        stack = [node]
        while(len(stack) > 0):
            node = stack.pop()
            if node.is_word:
                res += node.sum

            for k, v in node.children.iteritems():
                stack.append(v)
            
        return res     


# Your MapSum object will be instantiated and called as such:
# obj = MapSum()
# obj.insert(key,val)
# param_2 = obj.sum(prefix)

File: leetcode/677.map-sum-pairs.cpp
class TrieNode {
public:
    unordered_map<char, TrieNode *> children;
    bool isWord = false;
    int sum = 0;

    ~TrieNode(){
        for(auto it: children)
        {
            delete it.second;
        }
    }
};


class MapSum {
    TrieNode *root;
public:
    /** Initialize your data structure here. */
    ~MapSum() {
        delete root;
    }

    MapSum() {
        root = new TrieNode();
    }
    
    void insert(string key, int val) {
        TrieNode *node = root;
        for (auto c: key){
            if (node->children.count(c) == 0){
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
        }
        node->isWord = true;
        node->sum = val;
    }
    
    int sum(string prefix) {
        TrieNode *node = root;
        for (auto c: prefix){
            if (node->children.count(c) == 0){
                return 0;
            }

            node = node->children[c];
        }
        int sum = 0;

        queue<TrieNode *> q;
        q.push(node);

        while(!q.empty())
        {
            TrieNode *node = q.front();
            q.pop();

            if (node->isWord){
                sum += node->sum;
            }

            for(auto it: node->children){
                q.push(it.second); //:Fixme
            }
        }

        return sum;
    }
};

/**
 * Your MapSum object will be instantiated and called as such:
 * MapSum obj = new MapSum();
 * obj.insert(key,val);
 * int param_2 = obj.sum(prefix);
 */
